book/invalidation.md

```diff
@@ -167,6 +167,79 @@ API!
+
+[^other-reasons]: I'm sure there are all sorts of performance
+    improvements possible without implementing the invalidation
+    techniques from this chapter, but they are essential toward
+    achieving incrementalism, which is a kind of asymptotic guarantee
+    that simple optimizations won't achieve.
+
+If we want interactions to be fast, even on large, complex pages,
+we're going to need re-rendering the page to take time *proportional
+to the size of the change*, not proportional to the size of the page.
+This property is sometimes called *incrementality*, and it really
+constrains our browser. For example, incrementality means that during
+rendering, we can't even *traverse* the whole page, since that would
+take time proportional to the whole page, not the change being made.
+
+Now, note that when the page is first loaded, rendering is definitely
+going to take time proportional to the size of the page. So to achieve
+incrementality, we're going to need to think of the initial render and
+later re-renders differently. The initial render will create a cache
+that later renders invalidate and recompute parts of.[^big-change] If
+we're only invalidate a small part each time, this will only take time
+proportional to the changed portion of the page.
+
+[^big-change]: While initial and later renders are conceptually
+    different, they'll use the same code path. Basically, the initial
+    render will be one big change from no page to the initial page,
+    while later re-renders will incorporate smaller changes.
+
+In a real browser, every step of the rendering pipeline needs to be
+incremental, but this chapter focuses on layout. Layout is important
+and makes a good illustration of the idea. Incrementalizing layout
+means skipping the two most expensive parts of layout: building the
+layout tree and traversing it to compute layout fields. Ultimately,
+re-layouts will take under a millisecond when typing on this page.
+
+Getting there is going to be a bit of a rocky road, though. The whole
+idea is to only recompute what needs to be recomputed---but to know if
+something needs to be recomputed, we'll need to know if it depends on
+something that's changed. So the key will be tracking dependencies.
+Our `needs_style` and `needs_layout` flags already involve a bit of
+simple dependency reasoning, but to achieve incrementality we'll need
+to track much finer-grained dependencies, which means we'll need to
+track way, way more of them.
+
+So in fact most of this chapter is devoted to tracking dependencies,
+and the abstractions we can use to help us do it. We'll need to
+execute several large refactors of our layout engine, and only at the
+end will we see substantially faster layout. But the complexity is
+worth it---for the end user, to get a much more responsive browser,
+and for you, to see how real browsers make interactive pages fast.
+
+::: {.further}
+Implementing incremental layout is complex, but it's a core part of
```

Comment:
This could be rephrased to refer back to the "principle of incremental performance", and focus
on why it's important for browsers to provide this as a core value proposition to developers and users.
(Above I'm suggesting that we focus on the *fact* that we'll be focusing on that, but here we can talk about
why -- having to do with the role of a platform provider in a layered platform: platform engineer vs web developer vs user.